File: 25. Reverse Nodes in k-Group/25.go
func reverseKGroup(head *ListNode, k int) *ListNode {
	if head == nil {
		return nil
	}
	node := head
	for i := 0; i < k; i++ {
		if node == nil {
			return head
		}
		node = node.Next
	}
	newHead := reverse(head, node)
	head.Next = reverseKGroup(node, k)
	return newHead
}
func reverse(a *ListNode, b *ListNode) *ListNode {
	prev := b
	for a != b {
		next := a.Next
		a.Next = prev
		prev = a
		a = next
	}
	return prev
}

File: 25. Reverse Nodes in k-Group/25.py
class Solution:

    def reverseKGroup(self, head: Optional[ListNode],
                      k: int) -> Optional[ListNode]:
        if head is None:
            return None
        node = head
        for i in range(0, k):
            if node is None:
                return head
            node = node.next
        newHead = self.reverse(head, node)
        head.next = self.reverseKGroup(node, k)
        return newHead

    def reverse(self, a: Optional[ListNode],
                b: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        while a != b:
            next = a.next
            a.next = prev
            prev = a
            a = next
        return prev
